File: js/src/decimal.mjs
// Exact decimal arithmetic on native BigInt magnitudes. See spec/SPEC.md §4.
//
// A decimal is { neg, digits, scale }, meaning (neg ? -1 : 1) * digits / 10^scale.
// `digits` is the unscaled magnitude as a native BigInt (0n for zero).
// Zero is never negative. Scale is part of the value: 2.50 is digits 250n at
// scale 2, and stays "2.50" through addition.
// Carrying native BigInt mantissas eliminates intermediate string conversions
// during arithmetic operations, matching Python's int and SBCL's bignum cores.

import { fail } from './errors.mjs';

export const DIV_SCALE = 10;

// spec/SPEC.md §6.4. These bound the *value*; ROUND's scale cap and POWER's
// exponent cap bound *arguments*, and an argument cap is not a value cap —
// POWER's base is unbounded, so nesting one POWER inside another multiplies the
// exponents and steps straight over the exponent cap. Two independent numbers
// rather than one shared budget, because ROUND(99.5, 1000000) is 1 000 002
// digits and legal under the scale cap: a shared budget would have shrunk what
// the spec already sanctions.
export const MAX_INT_DIGITS = 1000000;
export const MAX_FRAC_DIGITS = 1000000;

const _MAX_INT_BITS = 3321929;
const _INT_LIMIT_SHIFT = BigInt(_MAX_INT_BITS - 1);
const _FAST_BOUND = 10n ** 18n;

const POW10_TABLE = [1n];
for (let i = 1; i <= 64; i++) POW10_TABLE.push(POW10_TABLE[i - 1] * 10n);
const POW10_CACHE = new Map();
const POW10_CACHE_ENTRIES = 64;
const POW10_CACHE_MAX_EXPONENT = 1000000;
const POW10_CACHE_DIGITS = 1048576;
let pow10Weight = 0;

export function pow10(k) {
  if (k <= 64) return POW10_TABLE[k];
  let v = POW10_CACHE.get(k);
  if (!v) {
    v = 10n ** BigInt(k);
    if (k <= POW10_CACHE_MAX_EXPONENT) {
      if (POW10_CACHE.size >= POW10_CACHE_ENTRIES || pow10Weight + k > POW10_CACHE_DIGITS) {
        POW10_CACHE.clear();
        pow10Weight = 0;
      }
      POW10_CACHE.set(k, v);
      pow10Weight += k;
    }
  }
  return v;
}

function bitLength(n) {
  if (n === 0n) return 0;
  const hex = n.toString(16);
  return (hex.length - 1) * 4 + (32 - Math.clz32(parseInt(hex[0], 16)));
}

function numDigits(n) {
  if (n === 0n) return 1;
  const bitLen = bitLength(n);
  let d = Math.floor((bitLen * 30103) / 100000) + 1;
  while (n < pow10(d - 1)) d--;
  return d;
}

function make(neg, digits, scale) {
  const d = typeof digits === 'bigint' ? digits : BigInt(digits);
  const actualNeg = d === 0n ? false : !!neg;
  return { neg: actualNeg, digits: d, scale };
}

// Refuses a value SEL cannot hold, where it is built rather than where it is
// rendered. Every operation that can grow a number passes its result through
// here, so POWER — repeated squaring over mul — trips on an intermediate and the
// enormous value is never allocated: without that, nesting POWER three deep
// exhausted the host's memory before any check could run.
function guard(d, pos) {
  if (d.scale > MAX_FRAC_DIGITS) {
    fail('E_RANGE', `number has more than ${MAX_FRAC_DIGITS} fractional digits`, pos);
  }
  if (d.digits > _FAST_BOUND) {
    // A shift past the magnitude returns zero without rendering its digits.
    // Only values near the cap need the exact digit count (and its hex string).
    if ((d.digits >> _INT_LIMIT_SHIFT) !== 0n && numDigits(d.digits) - d.scale > MAX_INT_DIGITS) {
      fail('E_RANGE', `number has more than ${MAX_INT_DIGITS} integer digits`, pos);
    }
  }
  return d;
}

export const ZERO = make(false, 0n, 0);

const NUM_RE = /^-?[0-9]+(\.[0-9]+)?$/;

// Returns null when the text is not a number; callers raise E_NOT_NUM with the
// position of the offending node. No trimming — " 2" is not a number.
//
// A well-formed numeral too big to hold is E_RANGE, not null: every character of
// it is a digit, so "not a number" would be false. Callers that must not raise —
// ISNUM's probe — catch it and answer no.
export function parse(text, pos) {
  if (typeof text !== 'string' || !NUM_RE.test(text)) return null;
  const neg = text.charCodeAt(0) === 45;
  const body = neg ? text.slice(1) : text;
  const dot = body.indexOf('.');
  const intPart = dot < 0 ? body : body.slice(0, dot);
  const fracPart = dot < 0 ? '' : body.slice(dot + 1);
  const stripped = (intPart + fracPart).replace(/^0+/, '') || '0';
  if (fracPart.length > MAX_FRAC_DIGITS) {
    fail('E_RANGE', `number has more than ${MAX_FRAC_DIGITS} fractional digits`, pos);
  }
  if (stripped.length - fracPart.length > MAX_INT_DIGITS) {
    fail('E_RANGE', `number has more than ${MAX_INT_DIGITS} integer digits`, pos);
  }
  return make(neg, BigInt(stripped), fracPart.length);
}

export function format(d) {
  const sign = d.neg ? '-' : '';
  const digitsStr = d.digits.toString();
  if (d.scale === 0) return sign + digitsStr;
  const padded = digitsStr.length <= d.scale
    ? '0'.repeat(d.scale - digitsStr.length + 1) + digitsStr
    : digitsStr;
  return sign + padded.slice(0, padded.length - d.scale) + '.' + padded.slice(padded.length - d.scale);
}

export function fromInt(n) {
  const big = BigInt(n);
  const neg = big < 0n;
  return make(neg, neg ? -big : big, 0);
}

export function isZero(d) {
  return d.digits === 0n;
}

export function negate(d) {
  return make(!d.neg, d.digits, d.scale);
}

export function abs(d) {
  return make(false, d.digits, d.scale);
}

export function sign(d) {
  return d.digits === 0n ? 0 : (d.neg ? -1 : 1);
}

// True when the value has no fractional part left after its scale is honoured.
export function isInteger(d) {
  if (d.scale === 0) return true;
  return d.digits % pow10(d.scale) === 0n;
}

export function toSafeInt(d) {
  const t = trunc(d);
  const v = Number(t.digits);
  return t.neg ? -v : v;
}

// --- arithmetic -------------------------------------------------------------

function aligned(a, b) {
  const s = Math.max(a.scale, b.scale);
  const A = a.scale === s ? a.digits : a.digits * pow10(s - a.scale);
  const B = b.scale === s ? b.digits : b.digits * pow10(s - b.scale);
  return [A, B, s];
}

export function add(a, b, pos) {
  const [A, B, s] = aligned(a, b);
  // Only true addition can grow: a difference is never wider than its operands,
  // and the aligned scale is the larger of two already legal ones.
  if (a.neg === b.neg) return guard(make(a.neg, A + B, s), pos);
  if (A === B) return make(false, 0n, s);
  return A > B ? make(a.neg, A - B, s) : make(b.neg, B - A, s);
}

export function sub(a, b, pos) {
  return add(a, negate(b), pos);
}

export function mul(a, b, pos) {
  return guard(make(a.neg !== b.neg, a.digits * b.digits, a.scale + b.scale), pos);
}

export function cmp(a, b) {
  if (a.digits === 0n && b.digits === 0n) return 0;
  if (a.neg !== b.neg) return a.neg ? -1 : 1;
  const [A, B] = aligned(a, b);
  const c = A === B ? 0 : (A < B ? -1 : 1);
  return a.neg ? -c : c;
}

// Exact when the quotient terminates within DIV_SCALE fractional digits (and
// then reported at its minimal scale); otherwise rounded half away from zero to
// exactly DIV_SCALE digits. So 4/2 is "2" and 1/3 is "0.3333333333".
export function div(a, b, pos) {
  if (b.digits === 0n) fail('E_DIV_ZERO', 'division by zero', pos);
  const N = a.digits * pow10(b.scale);
  const D = b.digits * pow10(a.scale);
  const num = N * 10000000000n; // pow10(DIV_SCALE)
  let q = num / D;
  const r = num % D;
  const neg = a.neg !== b.neg;
  if (r === 0n) {
    let digits = q;
    let scale = DIV_SCALE;
    while (scale > 0 && digits % 10n === 0n && digits !== 0n) {
      digits /= 10n;
      scale--;
    }
    if (digits === 0n) scale = 0;
    return guard(make(neg, digits, scale), pos);
  }
  if (2n * r >= D) q += 1n;
  return guard(make(neg, q, DIV_SCALE), pos);
}

// Remainder of truncated division: takes the sign of the dividend.
export function mod(a, b, pos) {
  if (b.digits === 0n) fail('E_DIV_ZERO', 'modulo by zero', pos);
  const [A, B, s] = aligned(a, b);
  return make(a.neg, A % B, s);
}

// --- rounding ---------------------------------------------------------------
//
// All of it half away from zero, on the magnitude, with the sign reattached by
// make(). Math.round is half-up (and floats besides) so it cannot appear here.

export function round(d, n, pos) {
  if (n >= d.scale) {
    return guard(make(d.neg, d.digits * pow10(n - d.scale), n), pos);
  }
  const p = pow10(d.scale - n);
  let q = d.digits / p;
  const r = d.digits % p;
  if (2n * r >= p) q += 1n;
  return guard(make(d.neg, q, n), pos);
}

export function trunc(d) {
  if (d.scale === 0) return d;
  return make(d.neg, d.digits / pow10(d.scale), 0);
}

export function floor(d) {
  if (d.scale === 0) return d;
  const p = pow10(d.scale);
  let q = d.digits / p;
  const r = d.digits % p;
  if (d.neg && r !== 0n) q += 1n;
  return make(d.neg, q, 0);
}

export function ceil(d) {
  if (d.scale === 0) return d;
  const p = pow10(d.scale);
  let q = d.digits / p;
  const r = d.digits % p;
  if (!d.neg && r !== 0n) q += 1n;
  return make(d.neg, q, 0);
}

// n must be a non-negative integer; the result scale is scale(x) * n, which
// falls out of repeated multiplication.
export function power(a, n, pos) {
  let result = make(false, 1n, 0);
  let base = a;
  // Arithmetic, not bit operators: JS's `&` and `>>` coerce to *32 bits*, so a
  // exponent above 2^31 silently wrapped and POWER(10, 4294967299) answered
  // 1000 with total confidence. Numbers are exact integers to 2^53 here.
  let e = n;
  while (e > 0) {
    if (e % 2 === 1) result = mul(result, base, pos);
    e = Math.floor(e / 2);
    if (e > 0) base = mul(base, base, pos);
  }
  return result;
}

File: python/sel/decimal.py
"""Exact decimal arithmetic. See spec/SPEC.md §4.

A decimal is (neg, digits, scale), meaning (neg ? -1 : 1) * digits / 10^scale.
`digits` is the unscaled magnitude, always non-negative. Zero is never negative.
Scale is part of the value: 2.50 is digits 250 at scale 2, and stays "2.50"
through addition.

**This deliberately does not use the `decimal` module.** Not for speed, and not
for pride: `tools/decimal-oracle.py` generates this project's decimal test cases
*from* Python's `decimal`, as an independent third opinion on four cores that
were all written from one spec by one hand. If this host's core were `decimal`
too, `tools/check-decimal.sh` would be comparing that module against itself and
would verify precisely nothing for Python while still printing a reassuring
"0 mismatches".

Where the other four hosts carry digit strings — because PHP has no bigint and
JS has doubles — this carries Python `int`, which is arbitrary precision. Every
digit-string routine there (addAbs, subAbs, mulAbs, divModAbs, scaleUp, strip)
is exactly an integer operation, so the mapping is one-to-one and the leading-
zero bookkeeping simply disappears.

The arithmetic is unbounded; the *conversions* are not. CPython refuses int/str
above 4300 digits by default, and since a SEL number is text, that limit reached
straight into the language: a 4301-digit literal the other four hosts evaluated
could not be compiled here at all. It is raised below to what §6.4 admits, and
the digit count the cap needs is read from bit_length rather than str(), because
str() is the conversion being guarded.
"""

from __future__ import annotations

from dataclasses import dataclass
import re
import sys

from .errors import Pos, fail

DIV_SCALE = 10

# spec/SPEC.md §6.4. These bound the *value*; ROUND's scale cap and POWER's
# exponent cap bound *arguments*, and an argument cap is not a value cap —
# POWER's base is unbounded, so nesting one POWER inside another multiplies the
# exponents and steps straight over the exponent cap. Two independent numbers
# rather than one shared budget, because ROUND(99.5, 1000000) is 1 000 002
# digits and legal under the scale cap: a shared budget would have shrunk what
# the spec already sanctions.
MAX_INT_DIGITS = 1000000
MAX_FRAC_DIGITS = 1000000

# The bit length at or above which a magnitude *may* have more than
# MAX_INT_DIGITS digits: floor(MAX_INT_DIGITS * log2(10)) + 1. Below it, it
# certainly does not. Used as an O(1) gate so the exact count is computed only
# for numbers that are actually near the cap.
_MAX_INT_BITS = 3321929

# CPython refuses int<->str conversion above 4300 digits by default — a guard
# against quadratic conversion, and a host policy, not SEL's answer. Under it a
# number this language admits could not be lexed or rendered at all: the Python
# host raised a ValueError out of the lexer for a 4301-digit literal that the
# other four hosts evaluated. Raised to exactly what SEL admits and no further.
#
# This mutates a process-global interpreter setting from inside a library, which
# is worth stating plainly. It is done in the one direction that cannot weaken a
# choice the application made: 0 means "unlimited" and is left alone, and a
# limit already above what SEL needs is left alone too.
_NEEDED_STR_DIGITS = MAX_INT_DIGITS + MAX_FRAC_DIGITS
if hasattr(sys, 'set_int_max_str_digits'):  # 3.11+
    _current = sys.get_int_max_str_digits()
    if _current != 0 and _current < _NEEDED_STR_DIGITS:
        sys.set_int_max_str_digits(_NEEDED_STR_DIGITS)


@dataclass(frozen=True, slots=True)
class Dec:
    neg: bool
    digits: int
    scale: int


def make(neg: bool, digits: int, scale: int) -> Dec:
    return Dec(False if digits == 0 else neg, digits, scale)


_POW10: dict[int, int] = {}


def _pow10(k: int) -> int:
    v = _POW10.get(k)
    if v is None:
        v = 10 ** k
        _POW10[k] = v
    return v


def _num_digits(n: int) -> int:
    """Digit count of a non-negative int, without str().

    str() is exactly what CPython refuses above its own limit, so a check that
    used it could not report on the values it exists to refuse. bit_length gives
    an upper bound (30103/100000 is just above log10(2)); one comparison walks
    it down to the exact count.
    """
    if n == 0:
        return 1
    d = (n.bit_length() * 30103) // 100000 + 1
    while n < _pow10(d - 1):
        d -= 1
    return d


def _guard(d: Dec, pos: Pos | None) -> Dec:
    """Refuses a value SEL cannot hold, where it is built rather than where it
    is rendered. Every operation that can grow a number passes its result
    through here, so power() — repeated squaring over mul() — trips on an
    intermediate and the enormous value is never allocated.

    The four hosts that store digits as a string read the count with strlen.
    Here it is an int, so the same question costs a bit_length gate and, only
    for numbers near the cap, an exact count.
    """
    if d.scale > MAX_FRAC_DIGITS:
        fail('E_RANGE', f'number has more than {MAX_FRAC_DIGITS} fractional digits', pos)
    # Negative when the value is below 1: those render as a single "0".
    if (d.digits.bit_length() >= _MAX_INT_BITS
            and _num_digits(d.digits) - d.scale > MAX_INT_DIGITS):
        fail('E_RANGE', f'number has more than {MAX_INT_DIGITS} integer digits', pos)
    return d


ZERO = make(False, 0, 0)

_NUM_RE = re.compile(r'^-?[0-9]+(\.[0-9]+)?$')
# NOTE ON .fullmatch(): Python's `$` matches at the end of the string *and*
# immediately before a single trailing newline, exactly like PCRE's and unlike
# ECMAScript's — the same trap the regex built-ins lower `$` to `\Z` for, one
# layer down and in this host's own source. With `.match()`, "5\n" parsed as a
# number here and as E_NOT_NUM everywhere else. `.fullmatch()` has no such
# corner; `$` is left in the pattern only because it costs nothing and reads.


def parse(text: str, pos: Pos | None = None) -> Dec | None:
    """None when the text is not a number; callers raise E_NOT_NUM with the
    position of the offending node. No trimming — " 2" is not a number.

    The regex is anchored and ASCII by construction. `int()` is never called on
    unvalidated text: it accepts Unicode digits ("٣" is 3), surrounding
    whitespace and underscore separators, any of which would let a value in that
    §4 does not define as a number.
    """
    if not isinstance(text, str) or not _NUM_RE.fullmatch(text):
        return None
    neg = text.startswith('-')
    body = text[1:] if neg else text
    dot = body.find('.')
    int_part = body if dot < 0 else body[:dot]
    frac_part = '' if dot < 0 else body[dot + 1:]
    # Measured before int(), because int() is the conversion CPython refuses on
    # a string this long. Leading zeros are stripped first: the string hosts
    # store strip(int_part + frac_part), so "0001" is one digit there and must
    # be one digit here.
    stripped = (int_part + frac_part).lstrip('0') or '0'
    if len(frac_part) > MAX_FRAC_DIGITS:
        fail('E_RANGE', f'number has more than {MAX_FRAC_DIGITS} fractional digits', pos)
    if len(stripped) - len(frac_part) > MAX_INT_DIGITS:
        fail('E_RANGE', f'number has more than {MAX_INT_DIGITS} integer digits', pos)
    return make(neg, int(stripped), len(frac_part))


def format(d: Dec) -> str:  # noqa: A001 - mirrors format() in the other hosts
    sign = '-' if d.neg else ''
    if d.scale == 0:
        return sign + str(d.digits)
    body = str(d.digits).rjust(d.scale + 1, '0')
    return sign + body[:len(body) - d.scale] + '.' + body[len(body) - d.scale:]


def from_int(n: int) -> Dec:
    return make(n < 0, abs(n), 0)


def is_zero(d: Dec) -> bool:
    return d.digits == 0


def negate(d: Dec) -> Dec:
    return make(not d.neg, d.digits, d.scale)


def abs_(d: Dec) -> Dec:
    return make(False, d.digits, d.scale)


def sign(d: Dec) -> int:
    return 0 if is_zero(d) else (-1 if d.neg else 1)


def is_integer(d: Dec) -> bool:
    """True when the value has no fractional part left after its scale."""
    if d.scale == 0:
        return True
    return d.digits % (10 ** d.scale) == 0


def to_safe_int(d: Dec) -> int:
    t = trunc(d)
    return -t.digits if t.neg else t.digits


# --- arithmetic -------------------------------------------------------------

def _aligned(a: Dec, b: Dec) -> tuple[int, int, int]:
    s = max(a.scale, b.scale)
    return a.digits * 10 ** (s - a.scale), b.digits * 10 ** (s - b.scale), s


def add(a: Dec, b: Dec, pos: Pos | None = None) -> Dec:
    A, B, s = _aligned(a, b)
    # Only true addition can grow: a difference is never wider than its
    # operands, and the aligned scale is the larger of two already legal ones.
    if a.neg == b.neg:
        return _guard(make(a.neg, A + B, s), pos)
    if A == B:
        return make(False, 0, s)
    return make(a.neg, A - B, s) if A > B else make(b.neg, B - A, s)


def sub(a: Dec, b: Dec, pos: Pos | None = None) -> Dec:
    return add(a, negate(b), pos)


def mul(a: Dec, b: Dec, pos: Pos | None = None) -> Dec:
    return _guard(make(a.neg != b.neg, a.digits * b.digits, a.scale + b.scale), pos)


def cmp(a: Dec, b: Dec) -> int:
    if is_zero(a) and is_zero(b):
        return 0
    if a.neg != b.neg:
        return -1 if a.neg else 1
    A, B, _ = _aligned(a, b)
    c = 0 if A == B else (-1 if A < B else 1)
    return -c if a.neg else c


def div(a: Dec, b: Dec, pos: Pos | None = None) -> Dec:
    """Exact when the quotient terminates within DIV_SCALE fractional digits
    (and then reported at its minimal scale); otherwise rounded half away from
    zero to exactly DIV_SCALE digits. So 4/2 is "2" and 1/3 is "0.3333333333".
    """
    if is_zero(b):
        fail('E_DIV_ZERO', 'division by zero', pos)
    N = a.digits * 10 ** b.scale
    D = b.digits * 10 ** a.scale
    q, r = divmod(N * 10 ** DIV_SCALE, D)
    neg = a.neg != b.neg

    if r == 0:
        digits, scale = q, DIV_SCALE
        while scale > 0 and digits % 10 == 0 and digits != 0:
            digits //= 10
            scale -= 1
        if digits == 0:
            scale = 0
        return _guard(make(neg, digits, scale), pos)
    return _guard(make(neg, q + 1 if 2 * r >= D else q, DIV_SCALE), pos)


def mod(a: Dec, b: Dec, pos: Pos | None = None) -> Dec:
    """Remainder of truncated division: takes the sign of the dividend."""
    if is_zero(b):
        fail('E_DIV_ZERO', 'modulo by zero', pos)
    A, B, s = _aligned(a, b)
    return make(a.neg, A % B, s)


# --- rounding ---------------------------------------------------------------
#
# All of it half away from zero, on the magnitude, with the sign reattached by
# make(). Python's own round() is half-to-even and must never appear here; nor
# may math.floor/ceil, which take floats.

def round(d: Dec, n: int, pos: Pos | None = None) -> Dec:  # noqa: A001 - mirrors round() in the other hosts
    if n >= d.scale:
        return _guard(make(d.neg, d.digits * 10 ** (n - d.scale), n), pos)
    p = 10 ** (d.scale - n)
    q, r = divmod(d.digits, p)
    # Rounding down still carries: 9.99 to one place is 10.0, a digit wider.
    return _guard(make(d.neg, q + 1 if 2 * r >= p else q, n), pos)


def trunc(d: Dec) -> Dec:
    if d.scale == 0:
        return d
    return make(d.neg, d.digits // (10 ** d.scale), 0)


def floor(d: Dec) -> Dec:
    if d.scale == 0:
        return d
    q, r = divmod(d.digits, 10 ** d.scale)
    return make(d.neg, q + 1 if d.neg and r != 0 else q, 0)


def ceil(d: Dec) -> Dec:
    if d.scale == 0:
        return d
    q, r = divmod(d.digits, 10 ** d.scale)
    return make(d.neg, q + 1 if not d.neg and r != 0 else q, 0)


def power(a: Dec, n: int, pos: Pos | None = None) -> Dec:
    """n must be a non-negative integer; the result scale is scale(x) * n, which
    falls out of repeated multiplication.
    """
    result = make(False, 1, 0)
    base = a
    e = n
    while e > 0:
        if e % 2 == 1:
            result = mul(result, base, pos)
        e //= 2
        if e > 0:
            base = mul(base, base, pos)
    return result
